cratetorrent/src/disk/io/piece.rs

```diff
@@ -119,18 +156,24 @@ impl Piece {
             // an empty file slice shouldn't occur as it would mean that piece
             // was thought to span fewer files than it actually does
             debug_assert!(file_slice.len > 0);
-            // the write buffer should still contain bytes to write
-            debug_assert!(!bufs.is_empty());
-            debug_assert!(!bufs[0].as_slice().is_empty());
 
-            // write to file
-            let tail = file.write(file_slice, bufs)?;
+            cfg_if! {
+                if #[cfg(target_os = "linux")] {
```

Comment:
I wonder if here instead of the two `cfg_if`s there could be a single call to a function that takes a closure which iterates over the files that the piece overlaps with and passes the file slice of each file to the given closure.

Something like:
```rust
cfg_if! {
    if #[cfg(target_os = "linux")] {
        // convert the blocks to IO slices that the underlying
        // pwritev syscall can deal with
        let mut blocks: Vec<_> = self
            .blocks
            .values()
            .map(|b| IoVec::from_slice(&b))
            .collect();
        // the actual slice of blocks being worked on
        let mut bufs = blocks.as_mut_slice();

        // write to the portion of the files that the piece overlaps with
        self.for_each_file_slice(files, |file_slice| {
            // the write buffer should still contain bytes to write
            debug_assert!(!bufs.is_empty());
            debug_assert!(!bufs[0].as_slice().is_empty());

            // write to file
            // `write_*` only writes at most `slice.len` bytes of
            // `bufs` to disk and returns the portion that wasn't
            // written, which we can use to set the write buffer for the next
            // round
            bufs = file.write_vectored(file_slice, bufs)?;
        })
    } else {
        // (the same situation is in the `read` method below)
        //
        // We don't have pwritev The Convenient. Instead, we have two options:
        //
        // 1. Coalesce the slices to a continuous preallocated memory buffer and
        //    write it at once.
        // 2. Iterate through the slices and issue a separate write call for each.
        //
        // The first option implies preallocating a short-living buffer
        // and destroying it shortly after. The second option
        // implies issuing a big number of writes, and possibly disk I/O operations.
        //
        // We've chosen the first option. Piece size is commonly less than 1MB
        // which is less than the page size, so the (re)allocation should be fast enough.
        // Careful benchmarking, while necessary when the project gets to the optimization phase,
        // is something I don't feel like doing just yet, and thus left as TODO
        // and an exercise to the reader.
        //
        // (another potential optimization is to preserve the buffer between `Piece::write` calls,
        // storing it in a `thread_local!` storage, but then the function will cease to be reentrant)
        let len = self.blocks.values().map(|s|s.len()).sum();
        let mut buffer = self
            .blocks
            .values()
            .fold(Vec::with_capacity(len), |mut buf, slice| {
                buf.extend_from_slice(slice);
                buf
            });
        let mut buffer = &buffer;

        // write to the portion of the files that the piece overlaps with
        self.for_each_file_slice(files, |file_slice| {
            debug_assert!(!buffer.is_empty());
            buffer = file.write(file_slice, buffer)?;
        })
    }
}
```

The same method could possibly be used for the read routine, since the concept is the same.

Reply:
I'll look into it in N days where N is *probably* less than 4. The new year was more eventful than expected (don't ask).

Reply:
No rush :) (same here btw)